### ISSUE TITLE
PyPy support

### ISSUE THREAD

**Issue by hawkinsp** [jax-ml/jax#474]
It would be interesting to combine JAX with PyPy, given we have `jit` to amortize the overhead of crossing the Python/C boundary.

I was able to successfully compile JAX against PyPy and CPyExt, but the result did not work due to an error in the swig bindings.

My gut feeling here: we should replace SWIG with a different binding framework. SWIG is pretty painful with or without PyPy. The obvious choices are `pybind11`, which TensorFlow may also move to (TBD), or something like `cffi` or `cppyy` which the PyPy authors recommend.

**Comment by mattjj** [comment]
Should we reach out to the PyPy devs for advice?

Anything tying us to the binding framework that TF uses (e.g. for internal or external build reasons)?

**Comment by hawkinsp** [comment]
No. There's nothing tying us to the same binding framework as TF. The only reason to prefer it is it might allow us to reuse TF infrastructure (e.g., build rules for SWIG, etc.).

**Comment by martinlightmatter** [comment]
FWIW I have had a lot of luck using `cffi` for a variety of other projects where I needed to enable C bindings. I have generally found them to be more robust and easy to use than other frameworks like SWIG. I don't have much experience to comment on `pybind11` vs `cppyy`, but have run into a few projects that struggled to tightly integrated `pybind11` with their usage of `setuptools`.

**Comment by hawkinsp** [comment]
Note we've already switched to pybind11 at Github head (although not in our released wheels).

So it's possible pypy might work now, given the SWIG/pypy problems are now history. At least, it's time for someone to experiment with it again.

`cffi` is a reasonable choice, but it would require us to define a C API to XLA, which, while possible, is a lot more work than simply binding the C++ API using pybind11.

**Comment by martinlightmatter** [comment]
I'll try to dig into this a little more over the weekend, but an initial pass at compiling in a `pypy3.6-7.1` environment failed with complaints from `jaxlib/lapack.pyx` such as:
```
  zheevd(&jobz, &uplo, &n, a_out, &n, w_out, work, &lwork, rwork, &lrwork,
                      ^
------------------------------------------------------------

jaxlib/lapack.pyx:788:23: Converting to Python object not allowed without gil
```

**Comment by mattjj** [comment]
That could be #375, maybe try updating your cython?

**Comment by hawkinsp** [comment]
It also may mean you don't have scipy installed.

**Comment by mattip** [comment]
I guess jax is not meant to be `pip install` -able from source by pypy? When I do `pypy3 -mpip install jax` I don't get `jaxlib`, and when I `pypy3 -mpip install jaxlib` and look at the result, `ls path/to/pypy3/site-packages/jaxlib` is missing a `version.py`. I think build succeeded, I see a `lapack.so`.

**Comment by mattjj** [comment]
Doing a `pip install jax` won't get you jaxlib in any case; see [the "pip installation" section of the install guide](https://github.com/google/jax#pip-installation).

I'm not sure about the missing version.py though. Do you get an error when trying to import it?

**Comment by mattip** [comment]
yes, I get an `ImportError`. Perhaps since I am downloading the source tarball from pypi and it is not being stored in that?

**Comment by superbobry** [comment]
I get the same error. I think this is due to `jaxlib` being a CPython extension (which PyPy could not load for some reason).

**Comment by sachinprasadhs** [comment]
Hi,

pip install jax now installs jax as well as jaxlib, 
Please find the below screenshot and the attached [Gist](https://gist.github.com/sachinprasadhs/c909e3093307431a844564b758fad18f) for reference, if your issue is resolved, could you please close the issue. 

![image](https://github.com/google/jax/assets/73069040/cc75c080-2ea7-4343-9291-c4a43d61f288)

**Comment by hawkinsp** [comment]
Nope, this issue pertains specifically to Pypy support, and JAX is unlikely to work on Pypy at present.

**Comment by hawkinsp** [comment]
Closing as not planned.
